<a href="https://colab.research.google.com/github/suwatoh/Python-learning/blob/main/129_%E7%92%B0%E5%A2%83%E3%81%A8%E3%82%B3%E3%83%B3%E3%83%95%E3%82%A3%E3%82%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

環境とコンフィグ
================

設定の優先順位
--------------

環境やユーザーの設定に応じてプログラムの動作が切り替わるようにすれば、プログラムはより広く使われるようになる。

プログラムがユーザーから値を受ける方法は何通りもあり、いくつかの方法に対応して優先順位を付けることが多く行われる。優先順位の付け方はプログラマーが自由に決めてよいが、一時的に使用される場所やアクセスしやすい場所の設定を優先するほうが直感的である。実際、多くの有名なプログラムでも、この方針で以下のような優先順位が採用されている:

  1. コマンドライン引数
  2. コマンドライン引数で明示的に指定された設定ファイル
  3. 環境変数
  4. カレントディレクトリにある設定ファイル
  5. ホームディレクトリまたは OS が推奨するアプリケーション用ディレクトリにある設定ファイル

環境に関する情報
----------------

### 実行中のプラットフォーム

`sys.platform` は、実行中のプラットフォーム（OS）を識別する文字列である（`sys.path` にプラットフォーム固有のサブディレクトリを追加することに利用されている）。

| システム | `platform` の値 |
|:---|:---|
| AIX | `'aix'` |
| Emscripten | `'emscripten'` |
| Linux | `'linux'` |
| WASI | `'wasi'` |
| Windows | `'win32'` |
| Windows/Cygwin | `'cygwin'` |
| macOS | `'darwin'` |

In [ ]:
import sys
sys.platform

'linux'

ちなみに、`os.name` は、Python 実装のコンパイル時に組み込まれたモジュールのうち、OS に依存するモジュールの名前（`'posix'`, `'nt'`, `'java'` のいずれか）である。

In [ ]:
import os
os.name

'posix'

OS の情報が場合分けに必要なわけではなく、単に OS の名前やバージョン表記を取得したいのであれば、`platform.system()`, `platform.release()`, `platform.version()` が使える。これらは不明な場合に空文字列を返すことに注意。

In [ ]:
import platform
platform.system(), platform.release(), platform.version()

('Linux', '6.1.85+', '#1 SMP PREEMPT_DYNAMIC Sun Apr 28 14:29:16 UTC 2024')

### コンピュータのネットワーク名

`platform.node()` は、コンピュータのネットワーク名を返す。これも不明な場合は空文字列を返す。

In [ ]:
import platform
platform.node()

'bcc35b63d48d'

### 実行中の Python インタープリター

`sys.version_info` は、Python のバージョン番号を示す 5 要素の名前付きタプル: `(major, minor, micro, releaselevel, serial)`。`releaselevel` 以外は全て整数である。

In [ ]:
import sys
sys.version_info

sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)

Python のバージョンを文字列で取得したいのであれば、`platform.python_version()` が使える。

In [ ]:
import platform
platform.python_version()

'3.10.12'

`sys.hexversion` は、単精度整数にエンコードされたバージョン番号。この値は `hex()` 関数を使って 16 進数に変換することで意味が明らかになる。たとえば、バージョン 3 以上であれば `0x03000000` 以上の値となる。バージョン 3.10 以上であれば `0x030a0000` 以上の値となる。バージョン 3.10.3 以上であれば `0x030a0300` 以上の値となる。

In [ ]:
import sys
hex(sys.hexversion)

'0x30a0cf0'

`sys.implementation` は、モジュール定数で、実行中の Python インタープリターに関する情報が格納されたオブジェクトとなる。以下の属性を持つ。

| 属性 | 意味 |
|:---|:---|
| `name` | 実装の識別子 |
| `version ` | `sys.version_info` と同じ |
| `hexversion` | `sys.hexversion` と同じ |
| `cache_tag` | キャッシュされたモジュールのファイル名に使用されるタグ |

In [ ]:
import sys
sys.implementation

namespace(name='cpython',
          cache_tag='cpython-310',
          version=sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0),
          hexversion=50990320,
          _multiarch='x86_64-linux-gnu')

`sys.flags` は、コマンドラインフラグの状態を表す名前付きタプル。各属性は読み出し専用になっている。属性名とフラグの対応は[公式ドキュメント](https://docs.python.org/ja/3/library/sys.html#sys.flags)を参照。

In [ ]:
import sys
sys.flags

sys.flags(debug=0, inspect=0, interactive=0, optimize=0, dont_write_bytecode=0, no_user_site=0, no_site=0, ignore_environment=0, verbose=0, bytes_warning=0, quiet=0, hash_randomization=1, isolated=0, dev_mode=False, utf8_mode=0, warn_default_encoding=0, int_max_str_digits=-1)

### psutil

サードパーティ製の [psutil](https://psutil.readthedocs.io/en/latest/) は、Python で実行中のプロセスとシステム（CPU、メモリ、ディスク、ネットワークなど）に関する情報を取得するためのクロスプラットフォームなライブラリである。ライセンスは BSD 3-clause License。インストール方法は次のとおり。

``` shell
pip install psutil
```

■ CPU 情報

In [ ]:
import psutil
print(psutil.cpu_count())               # CPUのコア数（論理）
print(psutil.cpu_count(logical=False))  # 物理コア数
print(psutil.cpu_percent(interval=1))   # CPU使用率（%）
print(psutil.cpu_times())               # ユーザー・システム・アイドル時間など

2
1
4.5
scputimes(user=237.14, nice=0.0, system=102.57, idle=2001.05, iowait=40.24, irq=0.0, softirq=5.71, steal=0.47, guest=0.0, guest_nice=0.0)


■ メモリ情報

In [ ]:
import psutil
mem = psutil.virtual_memory()
print(mem.total)     # 総メモリ
print(mem.used)      # 使用中メモリ
print(mem.free)      # 空きメモリ
print(mem.percent)   # 使用率（%）

13608361984
793387008
9820188672
8.0


■ ディスク情報

In [ ]:
import psutil
import pprint
pprint.pprint(psutil.disk_partitions())  # マウントポイント一覧
print(psutil.disk_usage('/'))            # 特定パスがあるパーティションのディスク使用状況統計
print(psutil.disk_io_counters())         # システム全体のディスクI/O統計

[sdiskpart(device='/dev/root', mountpoint='/usr/sbin/docker-init', fstype='ext2', opts='ro,relatime', maxfile=255, maxpath=4096),
 sdiskpart(device='/dev/sda1', mountpoint='/kaggle/input', fstype='ext4', opts='ro,relatime,commit=30', maxfile=255, maxpath=4096),
 sdiskpart(device='/dev/sda1', mountpoint='/etc/resolv.conf', fstype='ext4', opts='rw,nosuid,nodev,relatime,commit=30', maxfile=255, maxpath=4096),
 sdiskpart(device='/dev/sda1', mountpoint='/etc/hostname', fstype='ext4', opts='rw,nosuid,nodev,relatime,commit=30', maxfile=255, maxpath=4096),
 sdiskpart(device='/dev/sda1', mountpoint='/etc/hosts', fstype='ext4', opts='rw,nosuid,nodev,relatime,commit=30', maxfile=255, maxpath=4096)]
sdiskusage(total=115658190848, used=41868181504, free=73773232128, percent=36.2)
sdiskio(read_count=61175, write_count=70062, read_bytes=3077416960, write_bytes=837657600, read_time=52767, write_time=42522, read_merged_count=11415, write_merged_count=79002, busy_time=64267)


■ ネットワーク情報

In [ ]:
import psutil
import pprint
print(psutil.net_io_counters())       # 送受信バイト数など
pprint.pprint(psutil.net_if_addrs())  # ネットワークインターフェースのIPなど
pprint.pprint(psutil.net_if_stats())  # ネットワークインターフェースの状態

snetio(bytes_sent=3353016, bytes_recv=2700285, packets_sent=9862, packets_recv=11108, errin=0, errout=0, dropin=0, dropout=0)
{'eth0': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='172.28.0.12', netmask='255.255.0.0', broadcast='172.28.255.255', ptp=None),
          snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='02:42:ac:1c:00:0c', netmask=None, broadcast='ff:ff:ff:ff:ff:ff', ptp=None)],
 'lo': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='127.0.0.1', netmask='255.0.0.0', broadcast=None, ptp=None),
        snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='00:00:00:00:00:00', netmask=None, broadcast=None, ptp=None)]}
{'eth0': snicstats(isup=True, duplex=<NicDuplex.NIC_DUPLEX_FULL: 2>, speed=10000, mtu=1500, flags='up,broadcast,running,multicast'),
 'lo': snicstats(isup=True, duplex=<NicDuplex.NIC_DUPLEX_UNKNOWN: 0>, speed=0, mtu=65536, flags='up,loopback,running')}


■ プロセス管理

In [ ]:
import psutil
for proc in psutil.process_iter(['pid', 'name', 'username']):
    print(proc.info)  # PID・プロセス名・ユーザー名など
psutil.getloadavg()   # 過去 1 分間、5 分間、15 分間の CPU 待ちタスク数（実行中のタスクおよび待機中のタスクの平均数）

{'name': 'docker-init', 'pid': 1, 'username': 'root'}
{'name': 'node', 'pid': 7, 'username': 'root'}
{'name': 'oom_monitor.sh', 'pid': 21, 'username': 'root'}
{'name': 'run.sh', 'pid': 23, 'username': 'root'}
{'name': 'kernel_manager_proxy', 'pid': 25, 'username': 'root'}
{'name': 'tail', 'pid': 27, 'username': 'root'}
{'name': 'tail', 'pid': 33, 'username': 'root'}
{'name': 'python3', 'pid': 65, 'username': 'root'}
{'name': 'colab-fileshim.', 'pid': 66, 'username': 'root'}
{'name': 'jupyter-server', 'pid': 87, 'username': 'root'}
{'name': 'dap_multiplexer', 'pid': 88, 'username': 'root'}
{'name': 'python3', 'pid': 939, 'username': 'root'}
{'name': 'language_service', 'pid': 964, 'username': 'root'}
{'name': 'node', 'pid': 970, 'username': 'root'}
{'name': 'sleep', 'pid': 11841, 'username': 'root'}


(0.3310546875, 0.24755859375, 0.25244140625)

コマンドライン引数
------------------

### コマンドライン構文

以降は、Unix 系 OS でのコマンドライン構文である。Windows でも使える。

``` text
python script.py arg1 arg2 arg3
```

スペースで区切って並べた `arg1`、`arg2`、`arg3` は**位置引数**と呼ばれる。位置引数は並べる順番により識別される。位置引数のリストは可変長とすることもできる。

``` text
python script.py -o1 value1 -o2 --option3 value3
```

`-` を接頭辞として名前を書いた引数、および、`--` を接頭辞として名前を書いた引数は、**オプション引数**と呼ばれる。オプション引数は名前により識別される。スペースを置いて指定した値を格納するもの（`-o1` および `--option3`）と、フラグとするもの（`-o2`）がある。オプション引数は省略可能でなければならない。`-` を付けたほうを短縮名、`--` を付けたほうを正式名として 1 つのオプション引数と扱うことができる。たとえば、`-o1` を短縮名とし、`--option1` を正式名とするとき、コマンドライン上では `-o1` を指定しても `--option1` を指定しても同等とされる。

スクリプトの使い方を説明するヘルプメッセージを表示するオプション引数は、短縮名を `-h`、正式名を `--help` とする。

``` text
python script.py --option4 val1 val2 val3
```

オプション引数の後ろにスペースで区切って並べた `val1`、`val2`、`val3` は、リストとしてオプション引数 `--option4` の値とされる。

### argv

`sys.argv` は、Python スクリプトに渡されたコマンドライン引数を単純にスペースで分割した文字列のリストである。その最初の要素 `sys.argv[0]` は、Python インタープリターに渡すインターフェースオプションにより異なる。2 つの関係は、次のとおり。

| オプション | `sys.argv[0]` |
|:---|:---|
| `-c <command>` | `'-c'` |
| `-m <module-name>` | モジュールファイルのフルパス |
| `<python-file>` | `<python-file>` |
| `<directory>` | `<directory>` |
| `<zip-file>` | `<zip-file>` |
| `-` | `'-'` |

In [ ]:
import sys
sys.argv

['/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py',
 '-f',
 '/root/.local/share/jupyter/runtime/kernel-48dd642f-a43a-46ee-bf2b-4e61c5d0deac.json']

### argparse

引数を単純にリストで取得するのではなく、コマンドライン構文として解析するには、標準ライブラリの `argparse` モジュールを使用する。

`argparse` の基本的な使い方は、コマンドライン構文を解析するパーサーを `argparse.ArgumentParser` クラスのインストタンスとして作成し、その `add_argument()` メソッドでパーサーにコマンドライン引数の情報を与えることである。次は、もっとも簡単な例である。

``` python
import argparse
parser = argparse.ArgumentParser()
parser.parse_args()
```

下記がこのコードを実行した結果である:

``` text
$ python prog.py
$ python prog.py --help
usage: prog.py [-h]

options:
  -h, --help  show this help message and exit
$ python prog.py --foo
usage: foo.py [-h]
foo.py: error: unrecognized arguments: --foo
$ python prog.py foo
usage: prog.py [-h]
prog.py: error: unrecognized arguments: foo
```

  * コマンドライン引数なしの場合、何も起こらない。
  * コマンドライン引数 `--help`（または `-h`）がある場合、`add_argument()` で `--help` の情報が与えられていないにもかかわらず、スクリプトの使い方（ヘルプメッセージ）が標準出力に表示される。
  * `add_argument()` で情報が与えられていなかったコマンドライン引数が使用されている場合、引数が不正であることを示すエラーメッセージが標準エラー出力に表示される。

ヘルプメッセージもエラーメッセージも `argparse` が自動的に生成したものである。

また、`argparse.ArgumentParser` は `--help` が指定されると、ヘルプメッセージを表示した後 `sys.exit(0)` を呼び出す。そのため、`--help` を使うとプログラムは正常終了する。

例えば、`prog.py` ファイルが以下のようである場合:

``` python
import argparse
parser = argparse.ArgumentParser()
parser.parse_args()
print("スクリプトが続行されました")
```

このスクリプトを `python script.py --help` で実行すると、以下のような出力が表示され、その後スクリプトは終了する:

``` text
$ python test.py --help
usage: test.py [-h]

options:
  -h, --help  show this help message and exit
```

`'スクリプトが続行されました'` は表示されず、スクリプトは `sys.exit(0)` によって終了する。

``` python
argparse.ArgumentParser(prog=None, usage=None, description=None, epilog=None, parents=[], formatter_class=argparse.HelpFormatter,
                        prefix_chars='-', fromfile_prefix_chars=None, argument_default=None, conflict_handler='error', add_help=True,
                        allow_abbrev=True, exit_on_error=True)
```

`argparse.ArgumentParser` クラスのコンストラクタ引数は、主にヘルプメッセージを生成するために使用される。

| 引数 | 意味 |
|:---|:---|
| `prog` | プログラム名を指定する。`None` の場合は `os.path.basename(sys.argv[0])` が使われる |
| `usage` | プログラムの利用方法を文字列で指定する。`None` の場合はパーサーに追加された引数から生成される |
| `description` | 引数のヘルプの前に表示されるテキストを指定する。`None` の場合はテキストなし |
| `epilog` | 引数のヘルプの後に表示されるテキストを指定する。`None` の場合はテキストなし |
| `parents` | `ArgumentParser` オブジェクトのリストを指定する。このリストに含まれるオブジェクトの引数が追加される |
| `formatter_class` | ヘルプメッセージの整形をカスタマイズするためのクラスを指定する。デフォルトは `argparse.HelpFormatter` |
| `prefix_chars` | オプション引数の接頭辞になる文字集合を指定する。デフォルトは `'-'`。例えば `'+'` を指定すると `+o` のような引数になる |
| `fromfile_prefix_chars` | ファイルから追加の引数を読み込む場合に、そのファイルの名前の接頭辞になる文字列を指定する。`None`（デフォルト）の場合は接頭<br /><br />辞なし。例えば `'@'` を指定すると `@file.txt` のようにファイルを指定できる |
| `argument_default` | 全体的に適用される引数のデフォルト値を指定する。デフォルトは `None` |
| `conflict_handler` | すでに利用されているオプション文字列を使って新しい引数を作ろうとしたときの挙動を以下の文字列で指定する<br /><br />・`'error'`: `argparse.ArgumentError` 例外を送出する（デフォルト）<br /><br />・`'resolve'`: オプション文字列が上書きされる |
| `add_help` | `True`（デフォルト）の場合、`-h` 引数および `--help` 引数をパーサーに追加する |
| `allow_abbrev` | `True`（デフォルト）の場合、正式名の引数を先頭の 1 文字の短縮名で指定できるようにする |
| `exit_on_error` | `False` の場合、`--help` 実行時に `sys.exit()` せずに `argparse.ArgumentError` 例外を送出する（プログラム内で捕捉できる） |

以下は、`argparse.ArgumentParser` のメソッド。

``` python
ArgumentParser.parse_args(args=None, namespace=None)
```

このメソッドは、コマンドライン引数を解析し、`argparse.Namespace` オブジェクトを作成して返す。パーサーは、コマンドライン引数の解析結果を `argparse.Namespace` オブジェクトに属性を追加する形で保存する。

`parse_args()` メソッドの引数は、次のとおり。

| 引数 | 意味 |
|:---|:---|
| `args` | 解析する文字列のリストを指定する。`None`（デフォルト）なら、`sys.argv` から取得される |
| `namespace` | 戻り値のために使用する `argparse.Namespace` オブジェクトを指定する。`None`（デフォルト）なら、新しい空の `argparse.Namespace` オブジェクト<br /><br />が作成される |

``` python
ArgumentParser.add_argument(name or flags..., *[, action][, nargs][, const][, default][, type][, choices][, required][, help][, metavar][, dest][, deprecated])
```

このメソッドは、パーサーにコマンドライン引数の情報を与える。パーサーは、与えられた情報に基づいてコマンドライン引数を解析する。

`name or flags...` 以外はすべてキーワード引数である。

| 引数 | 意味 |
|:---|:---|
| `name or flags...` | `'foo'` のように位置引数の名前か、`'-f', '--foo'` のようにオプション引数の名前のリストを指定する |
| `dest` | `Namespace` に追加される属性の名前を指定する。省略する場合は、`name or flags...` が使われる |
| `action` | 引数をどのように処理するかを指定する<br /><br />・`'store'`: 引数に与えられた値を `Namespace` の属性に格納する（デフォルト）<br /><br />・`'store_const'`: 引数はフラグとなり、引数が指定されていれば `const` キーワード引数で指定された値を `Namespace` の属性に格納する<br /><br />・`'store_true'`: 引数はフラグとなり、引数が指定されていれば `True` を `Namespace` の属性に格納する。デフォルト値は `False` とする<br /><br />・`'store_false'`: 引数はフラグとなり、引数が指定されていれば `False` を `Namespace` の属性に格納する。デフォルト値は `True` とする<br /><br />・`'append'`: 引数は複数回使用可能で、それぞれ与えられた値を `Namespace` の属性値であるリストに追加する。たとえばコマンドライン引<br /><br />　数が `--foo 1 --foo 2` なら属性 `foo` が追加されて値が `['1', '2']` となる<br /><br />・`'append_const'`: 引数は複数回使用可能で、`const` キーワード引数で指定された値を `Namespace` の属性値であるリストに追加する<br /><br />・`'extend'`: 引数は複数回使用可能で、それぞれ与えられた値で `Na<br />mespace` の属性値であるリストを拡張する。たとえばコマンドラ<br /><br />　イン引数が `--foo f1 --foo f2 f3 f4` なら属性 `foo` が追加されて値が `['f1', 'f2', 'f3', 'f4']` となる<br /><br />・`'count'`: 引数の数を `Namespace` の属性に格納する。たとえば、名前を `'--verbose'` と `'-v'` とする引数についてコマンドライン引数<br /><br />　が `-vvv` なら属性 `verbose` が追加されて値が `3` となる<br /><br />・`'help'`: 引数はヘルプメッセージを表示し、終了する<br /><br />・`'version'`: 引数は `version` キーワード引数に指定したバージョン情報を表示して終了する<br /><br />`'help'`、`'version'` の場合には、`Namespace` に属性が追加されない |
| `nargs` | 指定する値によってコマンドライン引数が以下のように解釈される:<br /><br />・`N`（整数）: 引数の後ろの `N` 個の要素がコマンドラインから集められ、リストに格納される。たとえば `--foo` について `nargs=2` としコマン<br /><br />　ドライン引数が `--foo a b c` なら属性 `foo` が追加されて値が `['a', 'b']` となる（`c` は別の位置引数と解釈される）<br /><br />・`'?'`: 引数の後ろの 1 個の要素が値に使われるが、すぐ後ろが別のオプション引数ならデフォルト値が使われる<br /><br />・`'*'`: 引数の後ろのすべての要素がコマンドラインから集められ、リストに格納される<br /><br />・`'+'`: `'*'` と同様であるが、加えて、最低でも 1 つのコマンドライン引数が存在しない場合にエラーメッセージを生成する |
| `default` | コマンドラインで値が渡されなかった場合のデフォルト値を指定する |
| `type` | 引数に渡された値を指定した型に変換する。デフォルトでは `str`。ユーザーが定義した関数（や呼び出し可能オブジェクト）も使用できる |
| `choices` | 引数の値として許される値を格納したコンテナ型（`list` など）を指定する |
| `required` | `True` を指定すると、引数は省略できない（オプション引数のみ有効） |
| `help` | 引数の簡潔な説明を指定する |
| `metavar` | ヘルプメッセージの中で引数を参照するときに使われる名前を指定できる。省略する場合、デフォルトでは、位置引数には `dest` の値をそ<br /><br />のまま利用し、オプション引数については `dest` の値を大文字に変換して利用する |
| `deprecated` | Python 3.13 で追加。`True` の場合、引数が使用された際に、その引数が将来削除される予定であるとの警告が `sys.stderr` に出力される |

``` python
ArgumentParser.print_help(file=None)
```

このメソッドは、ヘルプメッセージを表示する。`file` が `None` の場合、 `sys.stdout` に出力される。

### 位置引数の利用

以降では、Colab 上で動作を確認するため、`parser.parse_args()` に引数のリストを渡している。ふつうにコマンドライン引数を解析するには、`parser.parse_args()` を引数なしで呼び出すこと。

以下は位置引数の簡単な例である。

In [ ]:
import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("echo")

    # python prog.py foo と同じ
    args = parser.parse_args(["foo"])
    print(f"echo 引数の値は {args.echo!r}")
    print("-" * 50)

    # python prog.py --help と同じ
    parser.print_help()

if __name__ == "__main__":
    main()

echo 引数の値は 'foo'
--------------------------------------------------
usage: colab_kernel_launcher.py [-h] echo

positional arguments:
  echo

options:
  -h, --help  show this help message and exit


コマンドラインで与えた位置引数 `foo` は、`parser.parse_args()` が返す `Namespace` オブジェクトの `echo` 属性に格納されることがわかる。ちなみに、コマンドラインで位置引数を与えないと、`error: the following arguments are required: echo` のようなエラーメッセージが表示されて終了する。

このヘルプメッセージは、`echo` が位置引数であることだけしかわからないものなので、`echo` の説明を加える:

In [ ]:
import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("echo", help="echo the string you use here")

    # python prog.py foo と同じ
    args = parser.parse_args(["foo"])
    print(f"echo 引数の値は {args.echo!r}")
    print("-" * 50)

    # python prog.py --help と同じ
    parser.print_help()

if __name__ == "__main__":
    main()

echo 引数の値は 'foo'
--------------------------------------------------
usage: colab_kernel_launcher.py [-h] echo

positional arguments:
  echo        echo the string you use here

options:
  -h, --help  show this help message and exit


デフォルトでは引数は文字列として格納されるので、他の型が必要な場合はパーサーに伝える必要がある:

In [ ]:
import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("square", help="display a square of a given number", type=int)

    # python prog.py 4 と同じ
    args = parser.parse_args(["4"])
    print(args.square**2)

if __name__ == "__main__":
    main()

16


このコードで` parser.parse_args(["four"])` （コマンドラインなら `python prog.py four`）のように `int` 型に変換できない引数を与えると、`error: argument square: invalid int value: 'four'` のようなエラーメッセージが表示されて終了する。

### オプション引数の利用

以下はオプション引数の例である。

In [ ]:
import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-o1", "--option1", required=True)
    parser.add_argument("-o2", action="store_true")
    parser.add_argument("--option3", type=int, choices=[1, 2, 3], default="adult")
    parser.add_argument("-o4", "--option4", nargs=2)

    # python prog.py -o1 foo --option3 3 -o4 spam eggs と同じ
    args = parser.parse_args(["-o1", "foo", "--option3", "3", "-o4", "spam", "eggs"])
    assert args.option1 == "foo"
    assert args.o2 is False  # デフォルト値
    assert args.option3 == 3
    assert args.option4 == ["spam", "eggs"]

if __name__ == "__main__":
    main()

`-o1` や `-o4` のように、短縮名と正式名の両方を指定することができ、その場合には正式名のほうで属性が追加されることに注意する。

`-o1` については、`required=True` を指定しているので省略できない。もしコマンドラインで省略されていた場合、`error: the following arguments are required: -o1/--option1` のようなエラーメッセージが表示されて終了する。

`--option3` については、`choices` 引数を指定しているので、リスト以外の要素（たとえば `4`）をコマンドラインで与えると、`error: argument --option3: invalid choice: 4 (choose from 1, 2, 3)` のようなエラーメッセージが表示されて終了する。

以下は、バージョンを表示するオプション引数の例である。

``` python
import argparse
parser = argparse.ArgumentParser(prog='PROG')
parser.add_argument('--version', action='version', version='%(prog)s 2.0')
```

下記がこのコードを実行した結果である:

``` text
$ python prog.py --version
PROG 2.0
```

### FileType オブジェクト

`argparse` では、 `ArgumentParser.add_argument()` の `type` 引数に渡すことができる `argparse.FileType` オブジェクトが利用できる。

``` python
argparse.FileType(mode='r', bufsize=-1, encoding=None, errors=None)
```

`argparse.FileType` オブジェクトは呼び出し可能であり、パスを渡して呼び出すとファイルオブジェクトを返す。コンストラクタ引数の意味は、`open()` 関数のものと同じ。

以下は、`argparse.FileType` の使用例である。

In [ ]:
import argparse
import sys

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("infile", nargs="?", type=argparse.FileType("r", encoding="UTF-8"), default=sys.stdin)
    parser.add_argument("outfile", nargs="?", type=argparse.FileType("w", encoding="UTF-8"), default=sys.stdout)

    # python prog.py sample_data/README.md と同じ
    args = parser.parse_args(["sample_data/README.md"])
    print(args.infile)
    print(args.outfile)

if __name__ == "__main__":
    main()

<_io.TextIOWrapper name='sample_data/README.md' mode='r' encoding='UTF-8'>


コマンドラインで与えた第 1 引数 `sample_data/README.md` は、そのパスから生成されるファイルオブジェクトに変換されて格納されることがわかる。これは、位置引数を文字列として格納し `open()` 関数に渡して開いてもほぼ同じことなのだが、存在しないパスを指定すると `FileNotFoundError` 例外が発生するのではなく、エラーメッセージが表示されて終了する点が異なる。つまり、`argparse.FileType` を使用する場合、`open()` 関数の呼び出しと例外処理のコードを書く必要がなくなる。

2 つの位置引数は、それぞれ `add_argument()` に `nargs='?` と `default=sys.stdin` を渡している。これにより、位置引数が省略可能で、デフォルトでは標準入力と標準出力が格納される。実際、コマンドラインで第 2 引数 `outfile` を与えていないので、標準出力（Colab 上では `ipykernel.iostream.OutStream`）が格納されている。これらの引数を指定していない場合は、位置引数を省略するとエラーメッセージが表示されて終了する。

`FileType` オブジェクトは擬似引数 `'-'` を識別し、読み込み用の `FileType` であれば `sys.stdin` を、書き込み用の `FileType` であれば `sys.stdout` に変換する:

In [ ]:
import argparse

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("infile", type=argparse.FileType(encoding="UTF-8"))

    # python prog.py - と同じ
    args = parser.parse_args(["-"])
    print(args.infile)

if __name__ == "__main__":
    main()

<_io.TextIOWrapper name='<stdin>' mode='r' encoding='utf-8'>


### サブコマンド

多くのプログラムは、その機能をサブコマンドへと分割する。たとえば、パッケージ管理システム `pip` は、`pip list`、`pip freeze`、`pip install` などのサブコマンドを利用でき、サブコマンドごとに異なるコマンドライン引数を必要とする。

`ArgumentParser` インスタンスは `add_subparsers()` メソッドによりサブコマンドをサポートしている。`add_subparsers()` メソッドは、特殊なアクションオブジェクトを返す。このオブジェクトには 1 つのメソッド `add_parser()` を持ち、コマンド名と `ArgumentParser` コンストラクタの任意の引数を受け取り、通常の方法で操作できる `ArgumentParser` オブジェクトを返す。

In [ ]:
import argparse

def main():
    # トップレベルのパーサーの作成
    parser = argparse.ArgumentParser(prog="PROG")
    subparsers = parser.add_subparsers()

    # "a" コマンドのためのパーサーを作成
    parser_a = subparsers.add_parser("a", help="a help")
    parser_a.add_argument("bar", type=int, help="bar help")

    # "b" コマンドのためのパーサーを作成
    parser_b = subparsers.add_parser("b", help="b help")
    parser_b.add_argument("--baz", choices="XYZ", help="baz help")

    # python prog.py a 12 と同じ
    args_a = parser.parse_args(["a", "12"])
    assert args_a.bar == 12

    # python prog.py b --baz Z と同じ
    args_b = parser.parse_args(["b", "--baz", "Z"])
    assert args_b.baz == "Z"

if __name__ == "__main__":
    main()

`python prog.py a --help` のようにサブコマンドごとにヘルプメッセージを出力させることもサポートされる。

`add_subparsers()` の引数:

| 引数 | 意味 |
|:---|:---|
| `title` | ヘルプ出力でのサブパーサーグループのタイトル。デフォルトは、`description` が指定されている場合は `'subcommands'` に、指定されていな<br /><br />い場合は位置引数のタイトルになる |
| `description` | ヘルプ出力に表示されるサブパーサーグループの説明。デフォルトは `None` |
| `prog` | サブコマンドのヘルプに表示される使用方法の説明。デフォルトではプログラム名と位置引数の後ろに、サブパーサーの引数が続く |
| `parser_class` | サブパーサーのインスタンスを作成するときに使用されるクラス。デフォルトでは現在のパーサーのクラス（例: `ArgumentParser`）になる |
| `action` | コマンドラインにこの引数があったときの基本のアクション |
| `dest` | サブコマンド名を格納する属性の名前。デフォルトは `None` で値は格納されない |
| `required` | サブコマンドが必須であるかどうかを指定する。デフォルトは `False` |
| `help` | ヘルプ出力に表示されるサブパーサーグループのヘルプ。デフォルトは `None` |
| `metavar` | 利用可能なサブコマンドをヘルプ内で表示するための文字列。デフォルトは `None` で、サブコマンドを `{cmd1, cmd2, ..}` のような形式で表す |

環境変数
--------

### 環境変数の基本

**環境変数**（environment variable）は、OS がプログラム（プロセス）に渡す設定情報を保持する仕組みである。外部からプログラムに値を与えることで、その挙動や設定を変更するために利用される。ほとんどの主要な OS（Linux、macOS、Windows など）は環境変数をサポートしている。

登録した環境変数は、その OS 上で動作するプロセスが参照する仕組みであり、基本的には他のユーザーや外部から直接見ることはできない。プログラムが利用する API キーなどの秘匿情報は、ソースコードに埋め込むよりは、環境変数に登録して参照する方が安全である。

環境変数は、親プロセスから子プロセスに複製されて継承される。子プロセスで環境変数を変更しない限り、内容は親プロセスと同一である。一方、子プロセスが自分の環境変数を変更しても、親プロセスには影響しない。

環境変数の名前は、通常英数字とアンダースコア `_` で構成される。慣例としてすべて大文字で記述し、単語の区切りにはアンダースコアを用いる。

環境変数には、システムによって定義されるものと、ユーザが任意に定義するものがある。

UNIX（Unix 系も含む）では、次のような代表的な環境変数がある。

| 環境変数 | 意味 |
|:---|:---|
| `LANG` | ロケール設定。日本語ロケールの場合は `ja_JP.UTF-8` などを指定する |
| `TZ` | タイムゾーン指定 |
| `PATH` | コマンド検索パスをコロン `:` で区切って指定する |
| `HOME` | ユーザのホームディレクトリ指定。`cd` コマンドで引数を指定しない場合、このディレクトリに移動する |
| `TMPDIR` | 一時ディレクトリ指定 |
| `USER` | ログインユーザー名を保持 |
| `PWD` | 現在の作業ディレクトリを保持 |
| `SHELL` | 現在のシェルの起動パスを保持 |

Windows では、次のような代表的な環境変数がある。

| 環境変数 | 意味 |
|:---|:---|
| `PATH` | コマンド検索パスをセミコロン `;` で区切って指定する |
| `HOMEDRIVE` | ホームフォルダのドライブ名を指定 |
| `HOMEPATH` | ホームフォルダのパス名を指定 |
| `APPDATA` | ユーザー設定やアプリケーションの構成ファイルなどを置くディレクトリを指定 |
| `LOCALAPPDATA` | キャッシュファイル、ダウンロード済みデータ、PC ごとに異なる一時的な設定などを置くディレクトリを指定 |
| `TEMP`/`TMP` | 一時ディレクトリ指定 |
| `USERNAME` | ログインユーザー名を保持 |

多くのプログラミング言語は環境変数を扱う標準的な方法を備えている。環境変数の値は特定のデータ型を持たないため、文字列として扱うのが一般的である。

Python では、標準ライブラリの os モジュールを利用して環境変数にアクセスできる。

``` python
os.environ
```

これは、現在のプロセスに設定されている環境変数を参照・操作することが可能なマッピングオブジェクトである。キーは環境変数名、値は文字列として保持され、辞書と同様に取得・追加・削除が可能である。

このマッピングの内容は、基本的には Python プロセスの起動時の環境変数を取り込んだものである。それ以後に変更された環境変数は `os.environ` を直接変更しない限り `os.environ` には反映されない。

`os.environ['VARNAME']='NEWVAL'` のように設定した値はそのプロセス内でのみ有効で、OS 全体に恒久的に登録されるわけではないが、子プロセスには引き継がれる。

存在しない環境変数を取得しようとすると、KeyError 例外が発生する。また、文字列以外の値を設定しようとすると、TypeError 例外が発生する。

In [14]:
import os
assert os.environ['LANG'] == 'en_US.UTF-8'
try:
    print(os.environ['FAKEVAR'])
except Exception as err:
    print(f"{type(err).__name__}: {err}")
try:
    os.environ['NEWVAR'] = 123
except Exception as err:
    print(f"{type(err).__name__}: {err}")

KeyError: 'FAKEVAR'
TypeError: str expected, not int


``` python
os.getenv(key, default=None)
```

この関数は、環境変数名 `key` の値を返す。戻り値は常に文字列である。名前 `key` を持つ環境変数が存在しないときは、この関数は `default` 引数に設定した値（`default` を設定しない場合は `None`）を返す。`getenv()` は `os.environ` を使用するため、インポート後に外部から行われた環境変数の変更を反映しないことに注意。

In [12]:
import os
os.environ['NEWVAR'] = '123'
assert os.getenv('NEWVAR') == '123'

### .env ファイル

アプリケーションや開発ツールは、しばしば起動時に `.env` ファイルを読み込み、そのプロセス内でのみ有効な環境変数を設定することがある。OS に直接環境変数を登録する方法では環境ごとに再設定が必要になるが、`.env` ファイルを利用すれば環境ごとに異なるファイルを用意して切り替えるだけで済む。

`.env` ファイルのファイル名は文字通り `.env` で、通常はプロジェクトのルートディレクトリに配置される。

`.env` ファイルの構文は以下の通り。

  * `キー=値` の形式で記述する。
  * 値にスペースを含める場合は引用符 `"` で囲む。
  * 行頭に `#` を置くことで、その行はコメントとして扱われる。

`.env` ファイルの例:

``` ini
# コメントは # で記述可能
API_KEY=your_api_key
APP_NAME="My Awesome App"
```

サードパーティ製の `python-dotenv` を使用すると、Python で `.env` ファイルを扱うことができる。ライセンスは BSD 3-clause License。インストール方法は次のとおり。

``` shell
pip install python-dotenv
```

``` python
dotenv.load_dotenv(dotenv_path=None, stream=None, verbose=False, override=False, interpolate=True, encoding="utf-8")
```

この関数は、`.env` ファイルに定義された環境変数を読み込み、`os.environ` に設定する。これにより、アプリケーション内で `os.getenv("KEY")` や `os.environ["KEY"]` を使って環境変数を参照できるようになる。

戻り値は、少なくとも 1 つの環境変数が設定されている場合は `True`、それ以外の場合は `False` を返す。

| 引数 | 意味 |
|:---|:---|
| `dotenv_path` | 読み込む `.env` ファイルのパスを指定する文字列、または `pathlib.Path` オブジェクト。指定しない場合はカレントディレクトリから親ディレクト<br /><br />リに向かって `.env` という名前のファイルを探す |
| `stream` | `.env` の内容を持つテキストストリーム（`io.StringIO` オブジェクトなど）。`dotenv_path` と `stream` は同時に指定できない |
| `verbose` | `True` の場合、読み込んだ環境変数や処理状況を標準出力に表示する |
| `override` | `True` の場合、既に `os.environ` に存在する変数を `.env` の値で上書きする。`False`（デフォルト）の場合は既存値を保持 |
| `interpolate` | `True`（デフォルト）の場合、`.env` 内で `${VAR}` のような変数参照を展開する。`False` の場合は展開しない |
| `encoding` | `.env` ファイルの文字コード。デフォルトは `"utf-8"` |

In [3]:
! echo "API_KEY=your_api_key" > /content/.env
from dotenv import load_dotenv
import os
load_dotenv()
print(os.getenv("API_KEY"))

your_api_key


設定ファイル
------------

### configparser

標準ライブラリの `configparser` モジュールは、INI ファイルを読み込む機能を提供する。

次は、INI ファイルの例である。

``` ini
[DEFAULT]
home_dir = /Users/User
limit = 100
debug = No

[Multiline Values]
chorus = I'm a lumberjack, and I'm okay
    I sleep all night and I work all day

# comment
; comment

[WORK]
data_dir = %(home_dir)s/AppData
```

`[]` で囲んだ行がセクションの始まりを表し、`[]` の中身がセクション名となる。セクションに `key = value` を書いて、キーと値を指定する。`=` の代わりに `:` を使ってもよい。値として複数行の文字列を指定する場合は、2行目以降を改行してインデントを付ければよい。

行の先頭に `#` または `;` を置くと、その行はコメントとして扱われる。

一度定義したキーを `%()s` で囲んで埋め込むと、キーの値を利用できる。

INI ファイルを読み込む手順は、以下のようになる。

  1. `configparser.ConfigParser()` コンストラクタの呼び出しで、パーサーを作成する。
  2. パーサーの `read(filename, encoding=None)` メソッドで、INI ファイルを読み込む。`filename` は path-like オブジェクト。
  3. パーサーに対して、辞書のように `[SECTION]` でセクションを参照するオブジェクトを得る。
  4. セクションオブジェクトに対して、辞書のように `[key]` で値を文字列として取得する。

``` python
import configparser
config = configparser.ConfigParser()
config.read("./config.ini", encoding="utf-8")
default = config["DEFAULT"]
home_dir = default["home_dir"]
# home_dir = config["DEFAULT"]["home_dir"] のような書き方もできる
```

値を数値として取得したい場合は、組み込み関数 `int()` または `float()` で型変換する必要がある。

値を `bool` 型で取得する場合は、パーサーのメソッド `getboolean(section)` が使える。このメソッドは、値の `'yes'` および `'true'`、`'on'`、`'1'` を `True` に変換し、`'no'` および `'false'`、`'off'`、`'0'` を `False` に変換する。

``` python
debug = config["DEFAULT"].getboolean("debug")
```

### tomllib

標準ライブラリの `tomllib` モジュールは、TOML を解析する機能を提供する（Python 3.11 で追加）。このモジュールは、TOML の書き出しをサポートしていない。

``` python
tomllib.load(fp, /, *, parse_float=float)
```

この関数は、TOMLファイルを読み込み、解析結果の辞書を返す。第 1 引数には読み込み可能なバイナリモードのファイルオブジェクトを指定する。TOML の型は以下の変換表を使用して Python のデータ型に変換される。

| TOML | Python |
|:---|:---|
| TOMLドキュメント | 辞書 |
| 文字列 | `str` |
| 整数 | `int` |
| 浮動小数点数 | `float` （`parse_float` で設定可能） |
| ブール値 | `bool` |
| オフセット付き日時 | `datetime.datetime`（`tzinfo` 属性は `datetime.timezone` のインスタンスが設定される） |
| ローカルの日時 | `datetime.datetime` （`tzinfo` 属性は `None` に設定される） |
| ローカルの日付 | `datetime.date` |
| ローカルの時刻 | `datetime.time` |
| 配列 | `list` |
| テーブル | `dict` |
| インラインテーブル | `dict` |
| テーブルの配列 | `list[dict]` |

TOML の浮動小数点数型に対しては、`parse_float` に指定された呼び出し可能オブジェクトが呼び出されて変換が行われる。たとえば `decimal.Decimal` を指定すれば 10 進数型への変換となる。

無効な TOML ドキュメントの場合は `tomllib.TOMLDecodeError` 例外が送出される。

TOML ファイルからデータを取得するコードは、次のようになる。

``` python
import tomllib
with open("pyproject.toml", "rb") as f:
    data = tomllib.load(f)
```

このように、バイナリモードでファイルオブジェクトを取得する必要がある。

``` python
tomllib.loads(s, /, *, parse_float=float)
```

この関数は、文字列 `s` から TOML を読み込むこと以外は、`load()` と同じである。

``` shell
>>> import tomllib
>>> toml_str = """
... python-version = "3.11.0"
... python-implementation = "CPython"
... """
>>> data = tomllib.loads(toml_str)
>>> data
{'python-version': '3.11.0', 'python-implementation': 'CPython'}
```

サードパーティ製の [Tomli-W](https://github.com/hukkin/tomli-w) パッケージは、 TOML の書き込み用に `tomllib` モジュールと組み合わせて使用でき、標準ライブラリの `pickle` モジュールと同様の書き込み API を提供する。ライセンスは MIT license。インストール方法は次のとおり。

``` shell
pip install tomli-w
```

文字列として保存:  
``` python
import tomli_w

doc = {"table": {"nested": {}, "val3": 3}, "val2": 2, "val1": 1}
expected_toml = """\
val2 = 2
val1 = 1

[table]
val3 = 3

[table.nested]
"""
assert tomli_w.dumps(doc) == expected_toml
```

ファイルへの書き込み:  
``` python
import tomli_w

doc = {"one": 1, "two": 2, "pi": 3}
with open("path_to_file/conf.toml", "wb") as f:
    tomli_w.dump(doc, f)
```

国際化
------

### 国際化と地域化

プログラムを世界の複数の地域に対応するようにしたい場合、必要に応じて機能的な変更や拡張を行うというやり方では、開発や保守に多くの時間と費用がかかる。

そこで、プログラムの本質的な機能と、複数の地域に対応する機能を分離し、プログラムに機能的な変更や拡張を加えることなく複数の地域に適合できるようにするための設計が考案された。そのような設計を**国際化**（internationalization）といい、 i18n と略される（internationalization の先頭の i と語尾の n の間に 18 文字があることに起因する）。主なものは以下の通り。

  * Unicode への対応
  * 右横書き言語への対応
  * 翻訳機能の組み込み
  * タイムゾーン対応の組み込み
  * 通貨情報の組み込み
  * 地域特有の表記への書式化

i18n の設計に基づいてプログラムを特定の地域に対応させるための実装は**地域化**（localization）と呼ばれ、L10N と略される（localization の先頭の l と語尾の n の間に 10 文字があることに起因する）。地域化には、縦書き、ルビへの対応など、地域固有の追加機能を含むこともある。

現代の OS は国際化されており、特定の地域への対応が簡単な設定だけで実現されている。例えば、Unix 系 OS の場合、環境変数などで言語の設定を行う。 Unix 系 OS 上で動く国際化されたプログラムは、環境変数などの設定を参照して地域の言語を決定している。

Python は言語レベルで Unicode に対応し、また、標準ライブラリの `datetime` モジュールでタイムゾーンに対応し、標準ライブラリの `locale` モジュールで通貨情報と地域特有の表記への書式化に対応している。

### 翻訳機能

クロスプラットフォームで開発されるプログラムでは、プログラム中の文字列を翻訳する機能に [GNU gettext](https://www.gnu.org/software/gettext/) ライブラリを利用することが広まっている。 GNU gettext では、以下のように翻訳機能をサポートする。

  1. ソースコードの中で翻訳対象とする文字列を `_("...")` で囲んでマークする。
  2. マークづけをしたファイルに `xgettext` コマンドを用いて翻訳可能な全ての文字列のリストを保持する .pot ファイルを生成する。
  3. 必要により .pot ファイルにコメントを付ける。コメントは翻訳に役立つ情報とする。ソースコード中の文字列の直前にコメントを置くと、そのコメントが .pot ファイルに引き継がれる。
  4. .pot ファイルを入力として `msginit` コマンドを実行し、言語コードの名前を持つ .po ファイルを生成する。日本語なら `ja.po`、韓国語なら `ko.po`、中国語（簡体字）なら `zh-CN.po`、スペイン語なら `es.po` というようになる。ファイル名を指定することも可能。
  5. 各 .po ファイルの中の文字列のリストに対して翻訳作業を行っていく。手作業またはツールにより .po ファイルを編集する。
  6. .poファイルを `msgfmt` コマンドで .mo バイナリファイルにコンパイルする。.mo ファイルは通常 `locale/<言語コード>_<国コード>/LC_MESSAGES` という言語ごとの専用のディレクトリに配置される。日本語なら `locale/ja_JP/LC_MESSAGES` である。プログラム本体と翻訳リソースをパッケージにして配布する。
  7. OS は環境変数より翻訳リソースを検索し（.mo ファイル中に該当言語のリソースがありさえすれば）、プログラムにその言語による表示を行わせることができる。

Python 標準ライブラリの `gettext` モジュールは、上記の過程 7 の機能に対する Python インターフェースを提供する。

``` python
gettext.translation(domain, localedir=None, languages=None, class_=None, fallback=False)
```

この関数は、`<localedir>/<language>/LC_MESSAGES/<domain>.mo` ファイルから翻訳リソースを探索する翻訳オブジェクトを返す。`<language>` はシーケンスである `languages` の要素である。`localedir` 引数が省略された場合、標準のシステムロケールディレクトリが使われる。`languages` 引数が省略された場合、環境変数 `LANGUAGE`、`LC_ALL`、`LC_MESSAGES`、および `LANG` が検索される。

`class_` 引数に翻訳オブジェクトの型を指定できる。指定されていない場合は `gettext.GNUTranslations` クラスとなる。

.mo ファイルが見つからなかった場合、 `fallback` 引数が `False`（デフォルト値）ならこの関数は `OSError` を送出し、`fallback` 引数が `True` なら `gettext.NullTranslations` インスタンスが返される。 `gettext.NullTranslations` クラスは、翻訳クラスを実装するのに使える基本的なインターフェースを提供している。 `gettext.GNUTranslations` クラスも `gettext.NullTranslations` のサブクラスである。 `gettext.NullTranslations` クラスのインターフェース自体は、未翻訳の文字列を出力する。

翻訳オブジェクトの主なメソッドは次のとおり。

| メソッド | 機能 | 戻り値 |
|:---|:---|:---|
| `gettext(message)` | `message` のローカライズされた翻訳を返す | `str` |
| `install(names=None)` | このオブジェクトの `gettext()` メソッドを組み込み名前空間 `__builtins__` にインストールし、変数 `_` に束縛する。<br /><br />`names` はインストールする別のメソッドをシーケンスで指定するための引数 | `None` |

翻訳オブジェクトの `install()` メソッドの使用は便利な方法ではあるが、組み込み名前空間 `__builtins__` にインストールするため、その影響はプログラム全体に及ぶことに注意する。モジュールのグローバルな名前空間にインストールするには、このメソッドの代わりに次のコードを使用して `_(...)` を使用できるようにする必要がある:

``` python
import gettext
t = gettext.translation('mymodule', ...)
_ = t.gettext
```

以下のファイル配置を持つプロジェクトで、`gettext` モジュールによる翻訳を実現する例を示す。

``` text
my-project/
    my_package/
        __init__.py
        __main__.py
    locale/
        ja_JP/
            LC_MESSAGES/
                messages.po
                messages.mo
        messages.pot
```

ここでは、`__main__.py` で翻訳機能を実装する。

`__main__.py`:  
``` python
import gettext
import os

import my_package

def init_translation():
    # 翻訳ファイルを配置するディレクトリ
    path_to_locale_dir = os.path.abspath(
        os.path.join(
            os.path.dirname(__file__),
            '../locale'
        )
    )

    # 翻訳用クラスの設定
    translater = gettext.translation(
        'messages',                   # domain: 辞書ファイルの名前
        localedir=path_to_locale_dir, # 辞書ファイル配置ディレクトリ
        languages=['ja_JP'],          # 翻訳に使用する言語
        fallback=True                 # .moファイルが見つからなかった時は未翻訳の文字列を出力
    )

    # gettext() メソッドを組み込み名前空間で _ に束縛する
    translater.install()

# プログラムを実行
if __name__ = '__main__':
    init_translation()
    my_package.main()
```

翻訳対象の文字列をマーク付けしておく。

`__init__.py`:  
``` python
def main():
    print(_('Hello, World!'))
```

.pot ファイルを生成する:  
``` shell
PS> xgettext -o locale/messages.pot my_package/__init__.py
```

.poファイルの作成:  
``` shell
PS> msginit -l ja_JP -o locale/ja_JP/LC_MESSAGES/messages.po -i locale/messages.pot
```

.poファイルを開いて、翻訳を入力する:  
``` text
msgid "Hello, World!"
msgstr "こんにちは、世界！"
```

.moファイルへのコンパイル:  

``` shell
PS> msgfmt locale/ja_JP/LC_MESSAGES/messages.po -o locale/ja_JP/LC_MESSAGES/messages.mo
```

すべてが設定されたら、プログラムを実行して、指定した言語の翻訳が適用されているか確認する。

``` shell
PS> python -m my_package
こんにちは、世界！
```